lxd/instance/drivers/driver_qemu.go

```diff
@@ -5716,7 +5716,49 @@ func (d *qemu) CanMigrate() (bool, bool) {
 
 // DeviceEventHandler handles events occurring on the instance's devices.
 func (d *qemu) DeviceEventHandler(runConf *deviceConfig.RunConfig) error {
-	return fmt.Errorf("DeviceEventHandler Not implemented")
+	if !d.IsRunning() {
+		return nil
+	}
+
+	if runConf == nil || len(runConf.Uevents) == 0 {
+		return nil
+	}
+
+	// Uevents will contain 1 entry at most, therefore we don't need to iterate through it.
+	for _, event := range runConf.Uevents[0] {
+		fields := strings.SplitN(event, "=", 2)
+
+		if fields[0] != "ACTION" {
+			continue
+		}
+
+		switch fields[1] {
+		case "add":
+			for _, usbDev := range runConf.USBDevice {
+				// This ensures that the device is actually removed from QEMU before adding it again.
+				// In most cases the device will already be removed, but it is possible that the
```

Comment:
How is it possible? Can you expand on this?

Reply:
I haven't tested it but in theory it should be possible. Here are the steps:
1) Run LXD
2) Launch VM
3) Attach USB device
4) Stop LXD (the VM continues to run)
5) Physically detach USB device

QEMU, in theory, should still have the USB device. If the device is plugged in again, and gets assigned the same devnum, QEMU will fail to create the device as it already exists.

Again, this is all theoretical :)

Reply:
Got you makes sense, so basically anything that causes LXD to miss the udev event?

Can you add a little comment there just explaining that for future readers? Thanks